### DISCUSSION TITLE
Strange behaviour of custom lexer

### DISCUSSION THREAD

**Discussion by gtribello** [pygments/pygments#2469] [General]
Hello

We are using pygments to do the syntax highlighting for a code that we maintain and we are getting a behaviour that we don't really understand.  We are not sure if this is a bug or a feature of pygments that we don't understand.  We have reproduced the behaviour in a minimal example.  If you could explain why this doesn't work we would really appreciate it.

So in the language we are highlighting every lines of input has a label and a name that we would like to colour in red and green respectively.  The label and name appear like this:

````
label: NAME 
````

We also have comments which are indicated using a #.  The colons can be trailing comments or whole line comments.  Our input can thus look like this:

````
# This is a comment
label: NAME      # This is also a comment
````

We have written the following lexer to differentiate between labels, names and comments:

```python 
class MiniPlumedLexer(RegexLexer):
   name = 'MiniPlumed'
   alisases = ['MiniPlumed']
   filenames = ['*.plmd']
   tokens = {
       'root': [
         # Deals with all comments
          (r'#.*$', Comment),
          # Find label: NAME
          (r'(.+?)(:\s+)(\S+\b)', bygroups(String, Text, Keyword))
       ]
   }
```

This works fine unless there is a : in the comment.  In other words, this input does not work:

````
# This is a comment
label: NAME      # This is also a comment with : a colon
````

If we render the above with the `HTMLFormatter` we get the following output:

```html
<div class="highlight"><pre><span></span><span class="c"># This is a comment </span>
<span class="s"> label</span>: <span class="k">NAME</span><span class="s">   # This is also a comment with </span>: <span class="k">a</span><span class="err"> colon</span>
</pre></div>
```

You can see that the part before the colon in the second comment is treated as if it is a label and the # appears to be ignored.  Our question is thus: why is the comment with the colon not found by the first regex we have defined in the Lexer?

My apologies if we are missing something obvious but I really have been tearing my hair out trying to understand this one.  Any help you can provide is greatly appreciated.

**Comment by jeanas** [comment]
A lexer works by starting at the beginning of the input and emitting tokens linearly. At the start of the input, the first regex doesn't match, since your input doesn't start with a `#`. The second regex, however, does match.

You probably want to change the second regex into `r'([^#]+?)(:\s+)([^\s#]+\b)`.

**Comment by gtribello** [comment]
Hello many thanks for responding to my question so quickly.  Also apologies that it has taken me so long to respond to your message.  

I just wanted to let you know that I managed to use your advice to fix the problem that we have been having.  

Thanks again for taking the time to respond.
